File: py/여행경로.py
def solution(tickets):
    # visit을 통해 이미 쓴 여행권인지 확인해야 함
    visit = {}
    for i, j in tickets:
        if (i, j) in visit:
            visit[i, j] += 1
        else:
            visit[i, j] = 1

    dic = {}
    for a, b in tickets:
        dic[a] = []
        dic[b] = []
    for a, b in tickets:
        dic[a].append(b)
    for i in dic:
        dic[i].sort()

    def check():
        for i in visit:
            if visit[i] != 0:
                return False
        return True

    global answer
    answer = []

    def dfs(now, value):
        global answer
        if answer:
            return
        if check():
            answer = value
            return
        for i in dic[now]:
            if visit[now, i] == 0:
                continue
            visit[now, i] -= 1
            dfs(i, value + [i])
            if answer:
                return
            visit[now, i] += 1

    dfs("ICN", ["ICN"])
    return answer

File: 프로그래머스/3/43164. 여행경로/여행경로.java
import java.util.*;
class Solution {
    static boolean[] visit;
    static HashMap<String,HashMap<Integer,String>> ticket = new HashMap<>();
    static Deque<String> answer = new LinkedList<>();
    static boolean find = false;
    static int count=0;
    
    public String[] solution(String[][] tickets) {
        visit = new boolean[tickets.length];
        for(int i=0;i<tickets.length;i++){
            String start = tickets[i][0];
            String end = tickets[i][1];
            if(!ticket.containsKey(start)) ticket.put(start,new HashMap<>());
            ticket.get(start).put(i,end);
        }
        answer.add("ICN");
        count=tickets.length;
        dfs("ICN",0);
        String[] ans = new String[answer.size()];
        int cc=0;
        while(!answer.isEmpty()) {ans[cc] = answer.pollFirst();cc++;}
        return ans;
    }
    
    public void dfs(String now,int num){
        if(num==count){ find=true; return;}
        HashMap<Integer,String> now_ticket = ticket.get(now);
        if(now_ticket == null) return; 
        List<Integer> keys = new ArrayList<>(now_ticket.keySet());
        keys.sort((a, b) -> now_ticket.get(a).compareTo(now_ticket.get(b)));
        
        for(int next : keys){
            if(visit[next]) continue;
            visit[next] = true;
            String move = now_ticket.get(next);
            answer.add(move);
            dfs(move,num+1);
            if(find) return;
            answer.pollLast();
            visit[next]=false;
        }
    }
}